extensions/jms/src/main/java/org/mule/extensions/jms/JmsSessionManager.java

```diff
@@ -0,0 +1,129 @@
+ * acknowledgement or a recover over a {@link Session}.
+ * This is used when the {@link AckMode} is configured in {@link AckMode#MANUAL}
+ *
+ * @since 4.0
+ */
+public class JmsSessionManager {
+
+  private static final Logger LOGGER = getLogger(JmsSessionManager.class);
+  private final Map<String, SessionInformation> pendingSessions = new HashMap<>();
+
+  /**
+   * Registers the {@link Message} to the {@link Session} using the {@code ackId} in order to being
+   * able later to perform a {@link AckMode#MANUAL} ACK
+   *
+   * @param ackId   the id associated to the {@link Session} used to create the {@link Message}
+   * @param message the {@link Message} to use for executing the {@link Message#acknowledge}
+   * @param jmsLock the optional {@link JmsListenerLock} to be able to unlock the {@link JmsListener}
+   * @throws IllegalArgumentException if no Session was registered with the given AckId
+   */
+  public void registerMessageForAck(String ackId, Message message, Session session, JmsListenerLock jmsLock) {
+    if (!pendingSessions.containsKey(ackId)) {
+      pendingSessions.put(ackId, new SessionInformation(message, session, jmsLock));
+    }
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Registered Message for Session AckId [" + ackId + "]");
+    }
+  }
+
+  /**
+   * Executes the {@link Message#acknowledge} on the latest {@link Message} associated to the {@link Session}
+   * identified by the {@code ackId}
+   *
+   * @param ackId the id associated to the {@link Session} that should be ACKed
+   * @throws JMSException if an error occurs during the ack
+   */
+  public void ack(String ackId) throws JMSException {
+    SessionInformation sessionInformation = getSessionInformation(ackId)
+        .orElseThrow(() -> new JmsAckException(format("No pending acknowledgement with ackId [%s] exists in this Connection",
+                                                      ackId)));
+
+    sessionInformation.getMessage().acknowledge();
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("Acknowledged Message for Session with AckId [" + ackId + "]");
+    }
+  }
+
+  /**
+   * Executes the {@link Session#recover()} over the {@link Session} identified by the {@code ackId}
+   *
+   * @param ackId the id associated to the {@link Session} used to create the {@link Message}
+   * @throws JMSException if an error occurs during recovering the session
+   */
+  public void recoverSession(String ackId) throws JMSException {
+    SessionInformation sessionInformation = getSessionInformation(ackId)
+        .orElseThrow(() -> new JmsSessionRecoverException("No pending session with ackId [" + ackId
+            + "] exists in this Connection"));
+
+    Session session = sessionInformation.getSession();
+
+    sessionInformation.getJmsListenerLock().ifPresent(lock -> {
+      if (lock.isLocked()) {
+        lock.unlock();
+      }
+    });
+
+    session.recover();
```

Comment:
also, we have to remove the recovered session from the "pending" sessions, since no message should be ack'd if it was already recovered... right? probably get's a little bit messy in the async mode